clientv3/integration/metrics_test.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/integration"
+	"github.com/coreos/etcd/pkg/testutil"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+func TestV3ClientMetrics(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	var (
+		addr string = "localhost:27989"
+		ln   net.Listener
+		err  error
+	)
+
+	// listen for all prometheus metrics
+	go func() {
+		mux := http.NewServeMux()
+		mux.Handle("/metrics", prometheus.Handler())
+
+		srv := &http.Server{Handler: mux}
+		ln, err = net.Listen("tcp", addr)
+
+		if err != nil {
+			t.Errorf("Error : %v occured while listening on addr : %v", err, addr)
+		}
+
+		err = srv.Serve(ln)
+
+		if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+			t.Fatalf("Err serving http requests: %v", err)
+		}
+	}()
+
+	url := "http://" + addr + "/metrics"
+
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	client := clus.Client(0)
+
+	w := clientv3.NewWatcher(client)
+	defer w.Close()
+
+	kv := clientv3.NewKV(client)
+
+	ctx := context.Background()
+
+	wBefore := sumCountersForMetricAndLabels(t, url, "grpc_client_msg_received_total", "Watch", "bidi_stream")
+	wc := w.Watch(ctx, "foo1")
+
+	pBefore := sumCountersForMetricAndLabels(t, url, "grpc_client_started_total", "Put", "unary")
+
+	_, err = kv.Put(ctx, "foo1", "bar")
+
+	if err != nil {
+		t.Errorf("Error putting value in key store")
+	}
+
+	pAfter := sumCountersForMetricAndLabels(t, url, "grpc_client_started_total", "Put", "unary")
+
+	if pBefore+1 != pAfter {
+		t.Errorf("grpc_client_msg_received_total expected : %d, got : %d", 1, pAfter-pBefore)
+	}
+
+	// consume watch response
+	select {
+	case <-wc:
+	case <-time.After(10 * time.Second):
+		t.Error("Timeout occured for getting watch response")
+	}
+
+	wAfter := sumCountersForMetricAndLabels(t, url, "grpc_client_started_total", "Watch", "bidi_stream")
+
+	if wBefore+1 != wAfter {
+		t.Errorf("grpc_client_msg_received_total expected : %d, got %d", 1, wAfter-wBefore)
+	}
+
+	ln.Close()
+}
+
+func sumCountersForMetricAndLabels(t *testing.T, url string, metricName string, matchingLabelValues ...string) int {
+	count := 0
+
+outer:
```

Comment:
@xiang90 if you see this its doing one check in fetchPrometheusLines and the other check here. I had not followed this because if we want all prometheus line we can re-use the 2nd method which i am calling. The method they are using is doing one level check where i am having no checks and the other where you pointed out.
Do you want it this way only?